### PR TITLE
Implementing custom role authorization

### PR DESCRIPTION
**Pull request by TheRealLenon**
Implementing the abilty for an custom role authentication including setting an default role for an user, which will prevent the authorization if the user has the default role.

### CODE REVIEW

File: src/KoIdentity/Model/UserRole.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright © Tekoding 2021. All rights reserved.
```
Comment (TekodingAdmin) [inline src/KoIdentity/Model/UserRole.cs:1]: Please add xml documentation

File: src/KoIdentity/Store/IUserRoleStore.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Copyright © Tekoding 2021. All rights reserved.
```
Comment (TekodingAdmin) [inline src/KoIdentity/Store/IUserRoleStore.cs:1]: Please add xml documentation

File: src/KoIdentity/Store/RoleStore.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
+// Copyright © Tekoding 2021. All rights reserved.
```
Comment (TekodingAdmin) [inline src/KoIdentity/Store/RoleStore.cs:1]: Please add xml documentation

File: src/KoIdentity/Store/UserRoleStore.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+// Copyright © Tekoding 2021. All rights reserved.
```
Comment (TekodingAdmin) [inline src/KoIdentity/Store/UserRoleStore.cs:1]: Please add xml documentation

File: src/KoIdentity/Authentication/JwtMiddleware.cs [hunk 5]
```diff
@@ -29,13 +30,13 @@ public JwtMiddleware(RequestDelegate nextRequest)
         /// The invocation of the <see cref="RequestDelegate"/> gets the JSON Web Token from the authorization header.
         /// </summary>
         /// <param name="context">The current <see cref="HttpContext"/>.</param>
-        public async Task Invoke(HttpContext context)
+        public async Task Invoke(HttpContext context, IUserManager userManager)
```
Comment (TekodingAdmin) [inline src/KoIdentity/Authentication/JwtMiddleware.cs:33]: Does this needs to be an Task?

File: src/KoIdentity/Authorization/AuthorizeAttribute.cs [hunk 6]
```diff
@@ -14,18 +16,45 @@ namespace Tekoding.KoIdentity.Authorization
...
+                return;
+            }
+
+            foreach (var currentRoleName in Roles)
+            {
+                if (userRoles == null || !userRoles.Exists(x => x.Role.Name == currentRoleName))
+                {
+                    context.Result = new JsonResult(new {message = "Unauthorized"})
+                        {StatusCode = StatusCodes.Status401Unauthorized};
+                }
```
Comment (TekodingAdmin) [inline src/KoIdentity/Authorization/AuthorizeAttribute.cs:57]: Please overthink the whole authorization login. can we make this more simple?

File: src/KoIdentity/Manager/IUserManager.cs [hunk 7]
```diff
@@ -69,12 +70,17 @@ public interface IUserManager<TUser, in TRegistrationDto, in TLoginDto>
         /// <typeparamref name="TUser"/>, if any, who has the specified <see cref="userId"/>.
         /// </returns>
         public Task<TUser> GetByIdAsync(Guid userId, CancellationToken cancellationToken = default);
+
+        public Task<List<TUserRole>> GetAllUserRolesByUserIdAsync(Guid userId,
```
Comment (TekodingAdmin) [inline src/KoIdentity/Manager/IUserManager.cs:74]: Please add xml documentation

File: src/KoIdentity/Manager/IUserManager.cs [hunk 8]
```diff
@@ -69,12 +70,17 @@ public interface IUserManager<TUser, in TRegistrationDto, in TLoginDto>
...
+            CancellationToken cancellationToken = default);
+
+        public Task<List<TRole>> GetAllRolesAsync();
```
Comment (TekodingAdmin) [inline src/KoIdentity/Manager/IUserManager.cs:77]: Please add xml documentation

**Review by TekodingAdmin — CHANGES_REQUESTED** [review]
Please do at least proper refactoring and adjust the xml documentation (so that no compiler warnings are seen)

File: src/KoIdentity/Manager/UserManager.cs [hunk 9]
```diff
@@ -148,6 +168,10 @@ public async Task<User> GetByIdAsync(Guid userId, CancellationToken cancellation
             return await UserStore.GetByIdAsync(userId, cancellationToken);
         }
 
+        public async Task<List<UserRole>> GetAllUserRolesByUserIdAsync(Guid userId,
+            CancellationToken cancellationToken = default) =>
+            await UserRoleStore.GetUserRolesByUserIdAsync(userId, cancellationToken);
```
[the remark this answers is not in the document]
Comment (TheRealLenon) [reply src/KoIdentity/Manager/UserManager.cs]: Created a new manager.

File: src/KoIdentity/Authentication/JwtMiddleware.cs (hunk 5 above)
Comment (TheRealLenon) [reply src/KoIdentity/Authentication/JwtMiddleware.cs]: Yes

File: src/KoIdentity/DatabaseContext.cs [hunk 10]
```diff
@@ -42,9 +42,23 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
             modelBuilder.Entity<User>().Property(u => u.MailAddress).IsRequired();
             modelBuilder.Entity<User>().Property(u => u.Password).IsRequired();
+            
+            modelBuilder.Entity<Role>().HasKey(r => r.Id);
+            modelBuilder.Entity<Role>().Property(r => r.CreationDate).HasDefaultValueSql("getDate()");
+            modelBuilder.Entity<Role>().Property(r => r.ChangeDate).HasDefaultValueSql("getDate()");
+            modelBuilder.Entity<Role>().Property(r => r.ChangeDate).ValueGeneratedOnAddOrUpdate();
+            modelBuilder.Entity<Role>().HasIndex(r => r.Name).IsUnique();
+            
+            modelBuilder.Entity<UserRole>().HasKey(ur => ur.Id);
```
[the remark this answers is not in the document]
Comment (TekodingAdmin) [reply src/KoIdentity/DatabaseContext.cs:52]: Will be considered in future updates.

File: src/KoIdentity/Authorization/AuthorizeAttribute.cs (hunk 6 above)
Comment (TekodingAdmin) [reply src/KoIdentity/Authorization/AuthorizeAttribute.cs:57]: Will be considered within future updates.

File: src/KoIdentity/Authentication/JwtMiddleware.cs [hunk 11]
```diff
@@ -47,10 +48,13 @@ public async Task Invoke(HttpContext context, IUserManager userManager, IUserRol
         /// Assigning the claims from the currently signed in user to the current context.
         /// </summary>
         /// <param name="context">The current <see cref="HttpContext"/>.</param>
+        /// <param name="userRoleManager">
+        /// The current <see cref="IUserRoleManager"/> used to access users in a persistence store.
+        /// </param>
         /// <param name="token">The token which contains the claims for the user.</param>
-        private static void AttachUserClaimsToContext(HttpContext context, IUserManager userManager, IUserRoleManager userRoleManager, string token)
+        private static void AttachUserClaimsToContext(HttpContext context, IUserRoleManager userRoleManager, string token)
```
Comment (TekodingAdmin) [inline src/KoIdentity/Authentication/JwtMiddleware.cs:55]: This can also be an Task

**Review by TekodingAdmin — CHANGES_REQUESTED** [review]

**Review by TekodingAdmin — APPROVED** [review]

**Pull request merged**